cdap-app-templates/cdap-program-report/src/main/java/co/cask/cdap/report/ReportGenerationSpark.java

```diff
@@ -389,6 +392,17 @@ private void generateReport(ReportGenerationRequest reportRequest, Location repo
       return null;
     }
 
+    /**
+     * @return a singleton executor for report generation
+     */
+    private static synchronized ExecutorService getReportGenerationExecutor() {
+      if (reportGenerationExecutor == null) {
+        reportGenerationExecutor =
+          Executors.newCachedThreadPool(Threads.createDaemonThreadFactory("report-generation"));
```

Comment:
Its still possible that this gets called multiple times (from different threads), also you didn't address terence's comment about it being daemon thread  and singleton executor.

Reply:
```
this.executor = new ThreadPoolExecutor(0, 1, 0, TimeUnit.SECONDS,
                                           new LinkedBlockingQueue<Runnable>(), new ThreadFactory() {
      @Override
      public Thread newThread(Runnable r) {
        Thread t = new Thread(null, r, "report-generation");
        t.setDaemon(true);
        return t;
      }
    }); 
```
you can something like above and doesn't have to handle shutdown since its a daemon thread, it will exit.

Reply:
I see, missed the method declaration, wouldn't it be enough to have the initialization in the constructor ? that should be called only once right ?

Reply:
Constructor will be invoke once per instance and there would be many instances, one per thread.

Reply:
I remember seeing multiple instances of the handler. Need to confirm whether the constructor will only be called once

Reply:
thanks terence, do we want to limit the max threads ?, currently the default is int_max for max threads.